File: 098_Validate_Binary_Search_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> pt;
    void traverse(TreeNode* root) {
        if (root->left) traverse(root->left);
        pt.push_back(root->val);
        if (root->right) traverse(root->right);
    }
    bool isValidBST(TreeNode* root) {
        if (root) traverse(root);
        for (int i = 1; i < pt.size(); i++) {
            if (pt[i] <= pt[i-1]) return false;
        }
        return true;
    }
};

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return isValidBST(root, NULL, NULL);
    }
    
    bool isValidBST(TreeNode* root, TreeNode* minNode, TreeNode* maxNode) {
        if (!root) return true;
        if (minNode && root->val <= minNode->val || maxNode && root->val >= maxNode->val) return false;
        return isValidBST(root->left, minNode, root) && isValidBST(root->right, root, maxNode);
    }
};

File: 098_Validate_Binary_Search_Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        def helper(root, min_node, max_node):
            if root is None:
                return True
            if (min_node and root.val <= min_node.val) or (max_node and root.val >= max_node.val):
                return False
            return helper(root.left, min_node, root) and helper(root.right, root, max_node)
        return helper(root, None, None)


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def inOrder(node, arr):
            if not node:
                return arr
            arr = inOrder(node.left, arr)
            arr.append(node.val)
            arr = inOrder(node.right, arr)
            return arr
        
        arr = inOrder(root, [])
        for i in range(1, len(arr)):
            if arr[i] <= arr[i - 1]:
                return False
        return True        
